### PR TITLE
Pokemon Emerald: Add Free Fly Blacklist

### PR DESCRIPTION
**Pull request by SansSkill**
## What is this fixing or adding?
Add an option to the yaml to blacklist certain locations from becoming the free fly location.

## How was this tested?
Generated multiple playthroughts blacklisting either all or all but one location.
Everytime all locations were blacklisted the free fly location defaulted to Littleroot Town.
Everytime all but one location was blacklisted the free fly location was set to the one location not blacklisted.

### CODE REVIEW

File: worlds/pokemon_emerald/__init__.py [hunk 1]
```diff
@@ -706,6 +706,7 @@ def fill_slot_data(self) -> Dict[str, Any]:
             "allowed_legendary_hunt_encounters",
             "extra_bumpy_slope",
             "free_fly_location",
+            "free_fly_blacklist",
```
Comment (Exempt-Medic) [inline worlds/pokemon_emerald/__init__.py:709]: Does this need to be in slot data?

Comment (SansSkill) [reply worlds/pokemon_emerald/__init__.py]: I don't think it does, thinking about it, so I've removed it

File: worlds/pokemon_emerald/options.py [hunk 2]
```diff
@@ -725,6 +725,25 @@ class FreeFlyLocation(Toggle):
     """
     display_name = "Free Fly Location"
 
+class FreeFlyBlacklist(OptionSet):
+    """
+    Disables specific locations as valid free fly locations.
+    Does nothing if Free Fly Location is set to false.
+    If all locations are disabled free fly location will be set to Littleroot Town.
```
Comment (Zunawe) [inline worlds/pokemon_emerald/options.py:732]:
```suggestion
class FreeFlyBlacklist(OptionSet):
    """
    Disables specific locations as valid free fly locations.
    Does nothing if Free Fly Location is disabled.
```
I don't think this line is necessary for players. I think it's reasonable to assume that if there is no enabled free fly location, then the option effectively does nothing, which is what setting it to Littleroot does. In fact, it's ultimately identical on the mod side as well, because it just sets the city to "visited", which triggers for Littleroot as soon as you step out of the truck anyway. And the spoiler won't show free fly in the playthrough if it's not logically relevant.

Also, "set to false" is a very text editor way to word the second line. I think "disabled" might be more consistent with other options.

File: worlds/pokemon_emerald/locations.py [hunk 3]
```diff
@@ -173,18 +173,22 @@ def set_free_fly(world: "PokemonEmeraldWorld") -> None:
...
+            "EVENT_VISITED_SLATEPORT_CITY" if "Slateport City" not in world.options.free_fly_blacklist else None,
+            "EVENT_VISITED_MAUVILLE_CITY" if "Mauville City" not in world.options.free_fly_blacklist else None,
+            "EVENT_VISITED_VERDANTURF_TOWN" if "Verdanturf Town" not in world.options.free_fly_blacklist else None,
+            "EVENT_VISITED_FALLARBOR_TOWN" if "Fallarbor Town" not in world.options.free_fly_blacklist else None,
+            "EVENT_VISITED_LAVARIDGE_TOWN" if "Lavaridge Town" not in world.options.free_fly_blacklist else None,
+            "EVENT_VISITED_FORTREE_CITY" if "Fortree City" not in world.options.free_fly_blacklist else None,
+            "EVENT_VISITED_LILYCOVE_CITY" if "Lilycove City" not in world.options.free_fly_blacklist else None,
+            "EVENT_VISITED_MOSSDEEP_CITY" if "Mossdeep City" not in world.options.free_fly_blacklist else None,
+            "EVENT_VISITED_SOOTOPOLIS_CITY" if "Sootopolis City" not in world.options.free_fly_blacklist else None,
+            "EVENT_VISITED_EVER_GRANDE_CITY" if "Ever Grande City" not in world.options.free_fly_blacklist else None,
```
Comment (Zunawe) [inline worlds/pokemon_emerald/locations.py:186]: I think it would be better to build a dict from the option keys to event names and then turn the option value into a list. Simplifies a little and looks nicer.
```py
my_dict = {
    "Slateport City": "EVENT_VISITED_SLATEPORT_CITY",
    # etc.
}

...

free_fly_locations = sorted(set(my_dict.values) - set(my_dict[city] for city in world.options.free_fly_blacklist.value))
fly_location_name = world.random.choice(free_fly_locations )
```

Something to that effect, I'm writing this entirely in the GH text editor, so I'm not going to guarantee correctness. Make sure to keep the `sorted` for determinism. Otherwise, iterating over a set isn't deterministic.

It'll need a _little_ more code to account for `world.options.free_fly_blacklist.value` containing every possible city, otherwise `random.choice` will complain at you for giving it an empty list.

**Review by Zunawe — COMMENTED** [review]
A couple overall comments:

- As discussed in Discord, we might as well expand this to all cities and set the default to exclude the sphere 1 cities. And with that change, we should update the option description for free fly location to remove the mention of sphere 1 cities, since that would be handled by this new option.
- I'm wondering if it's better for this to be a whitelist or blacklist. I think a whitelist would be slightly cleaner in the code, and a blacklist might be inverted from people's mental model. You have to think to yourself "I'd be okay going here, here, or here, so let me enable the blacklist for every location that isn't those three"
- Could you add a line to the changelog file under features for 2.3.0? Just in case it gets merged. We can move it later in this PR if it doesn't make the next release.

File: worlds/pokemon_emerald/locations.py [hunk 4]
```diff
@@ -173,18 +185,9 @@ def set_free_fly(world: "PokemonEmeraldWorld") -> None:
...
-            "EVENT_VISITED_LAVARIDGE_TOWN",
-            "EVENT_VISITED_FORTREE_CITY",
-            "EVENT_VISITED_LILYCOVE_CITY",
-            "EVENT_VISITED_MOSSDEEP_CITY",
-            "EVENT_VISITED_SOOTOPOLIS_CITY",
-            "EVENT_VISITED_EVER_GRANDE_CITY",
-        ])
+        free_fly_locations = sorted(map(lambda city: BLACKLIST_OPTION_TO_VISITED_EVENT[city], world.options.free_fly_blacklist.value))
+        if free_fly_locations:
+            fly_location_name = world.random.choice(free_fly_locations)
```
Comment (Zunawe) [inline worlds/pokemon_emerald/locations.py:190]: This is inverted, isn't it? For this code, `free_fly_locations` contains the event names for the cities that the user marked (blacklisted). So we want to choose from the full list minus the blacklist. That's why I had `set` operations in that comment. And the `sorted` only needs to be applied because it turns a set back into a list with deterministic order (no need to sort the `map`, in other words).

For the record, the community supported the blacklist version of this option, so we're going with that even though I suggested a whitelist earlier.

Comment (Zunawe) [reply worlds/pokemon_emerald/locations.py:190]: And while I'm here, a comprehension/generator looks better than `map` to me.
```py
set(BLACKLIST_OPTION_TO_VISITED_EVENT[city] for city in world.options.free_fly_blacklist.value)
```
This creates a set of event-named strings from the player's option. And then you can subtract it from the set of all of them. And then make a sorted list of the result.

**Review by Zunawe — APPROVED** [review]
Awesome. LGTM, ran it in the debugger to make sure it behaved correctly with a full and partial blacklist

**Comment by Zunawe** [comment]
I knew I forgot something when I clicked approve.

Please _don't_ add any lines to the changelog. I wanted it there while we were in an RC phase in case the PR got merged so that I wouldn't have to make a separate PR. But when we're not in an RC phase, modifying the changelog just causes conflicts and has the potential to make Archipelago's release notes confusing, since for the Emerald changes they link directly to the file on `main` (since the tag doesn't exist until after the release is made).

tl;dr, don't worry about the changelog any more. I'll fix it up when I need to.

**Review by nicholassaylor — APPROVED** [review]

**Comment by Exempt-Medic** [comment]
So y'all settled on a blacklist approach instead of a whitelist one?

**Comment by Exempt-Medic** [comment]
And also not doing this?
> As discussed in Discord, we might as well expand this to all cities and set the default to exclude the sphere 1 cities. And with that change, we should update the option description for free fly location to remove the mention of sphere 1 cities, since that would be handled by this new option.

**Comment by Zunawe** [comment]
Community voted to use a blacklist.

Forgot about the expanding to include sphere 1 cities. I'll take it either way, I think. I can add them later.

File: worlds/pokemon_emerald/locations.py [hunk 5]
```diff
@@ -173,18 +185,10 @@ def set_free_fly(world: "PokemonEmeraldWorld") -> None:
...
-            "EVENT_VISITED_FALLARBOR_TOWN",
-            "EVENT_VISITED_LAVARIDGE_TOWN",
-            "EVENT_VISITED_FORTREE_CITY",
-            "EVENT_VISITED_LILYCOVE_CITY",
-            "EVENT_VISITED_MOSSDEEP_CITY",
-            "EVENT_VISITED_SOOTOPOLIS_CITY",
-            "EVENT_VISITED_EVER_GRANDE_CITY",
-        ])
+        blacklisted_locations = set(BLACKLIST_OPTION_TO_VISITED_EVENT[city] for city in world.options.free_fly_blacklist.value)
+        free_fly_locations = sorted(set(BLACKLIST_OPTION_TO_VISITED_EVENT.values()) - blacklisted_locations)
```
Comment (Exempt-Medic) [inline worlds/pokemon_emerald/locations.py:189]: Couldn't you define `set(BLACKLIST_OPTION_TO_VISITED_EVENT.values()` at the same time you do `BLACKLIST_OPTION_TO_VISITED_EVENT` as something like `ALL_CITIES` ?

Comment (Exempt-Medic) [reply worlds/pokemon_emerald/locations.py:189]: I'm just gonna merge this and leave resolutions of this and earlier discussion to @Zunawe

**Review by Exempt-Medic — APPROVED** [review]
